android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java

```diff
@@ -1093,7 +1093,15 @@ private void dispatchCallback(String name, KrollDict data) {
 
 		data.put("source", activityProxy);
 
-		activityProxy.callPropertyAsync(name, new Object[] { data });
+		// TIMOB-19903
+		if (TiApplication.getInstance().runOnMainThread()) {
+			// We must call this synchornously to ensure it happens before we release the Activity reference on the V8/Native side!
+			activityProxy.callPropertySync(name, new Object[] { data });
+		} else {
+			// This hopefully finishes before we release the reference on the native side?! I have seen it crash because it didn't before though...
+			// Not sure it's safe to keep this behavior...
+			activityProxy.callPropertyAsync(name, new Object[] { data });
```

Comment:
I'm not sure I follow. With recent changes (required by hyperloop) the KrollRuntime thread may be the same as the UI thread or separate depending on a tiapp property.

callPropertyAsync always sticks a message in the queue to fire the JS method. When the KrollRuntime thread is the main/UI thread, that means it's guaranteed not to fire until after all the work it's currently doing - so it won't ever run until after we've already destroyed the activity proxy during onDestroy flows.

When there are separate main and UI threads, I'm not sure that there's any guarantee of ordering between the onDestroy/onStop callback and the release of the activity proxy. So I think it will effectively be a race condition. I have anecdotally seen this when debugging and not running on the main thread - where closing a window would cause a crash. But perhaps I'm wrong and we can guarantee/know that the on* callback will always be fired and finished before the release call later in the onDestroy flow?